### PR TITLE
Fix for Unicode compliance in Python 2.7 

### LINKED ISSUE #16094: Unicode compliance in Python 2.7 - string_types instead of str
Inspired by #16093 

`str` does not cover unicode strings (available in Python 3). In sympy this is solved by the compatibility type string_types.
https://github.com/sympy/sympy/blob/d44e6038bec476750bcbaf0fcffff4a108097e3e/sympy/core/compatibility.py#L22-L23

Still, quite a lot of code compares to `str` which may disallow unicode strings.

```
bin/sympy_time_cache.py:66:        if isinstance(f, str):
doc/ext/docscrape_sphinx.py:197:            if isinstance(self['References'], str):
sympy/combinatorics/fp_groups.py:1138:        elif isinstance(C.P[i][j], str):
sympy/core/logic.py:374:        if isinstance(arg, str):
sympy/core/multidimensional.py:22:    elif isinstance(n, str):
sympy/core/multidimensional.py:97:            if not isinstance(a, (int, str)):
sympy/core/multidimensional.py:122:                elif isinstance(n, str):
sympy/holonomic/holonomic.py:116:            if isinstance(generator, str):
sympy/holonomic/recurrence.py:71:            if isinstance(generator, str):
sympy/integrals/rubi/parsetools/parse.py:244:        if isinstance(i, str):
sympy/physics/mechanics/particle.py:42:        if not isinstance(name, str):
sympy/physics/mechanics/rigidbody.py:52:        if not isinstance(name, str):
sympy/physics/vector/frame.py:213:        if not isinstance(ind, str):
sympy/physics/vector/point.py:208:        if not isinstance(name, str):
sympy/plotting/experimental_lambdify.py:558:        if isinstance(tree, str):
sympy/plotting/experimental_lambdify.py:569:        if isinstance(tree, str):
sympy/plotting/pygletplot/color_scheme.py:57:        elif len(args) == 1 and isinstance(args[0], str):
sympy/plotting/pygletplot/plot_interval.py:25:            elif isinstance(args[0], str):
sympy/plotting/pygletplot/plot_mode.py:147:        if isinstance(mode_arg, str):
sympy/plotting/pygletplot/plot_mode.py:379:            if isinstance(a, str):
sympy/plotting/pygletplot/util.py:167:    if not isinstance(s, str):
sympy/polys/orderings.py:228:    if isinstance(order, str):
sympy/polys/polyoptions.py:312:        elif isinstance(wrt, str):
sympy/polys/polyoptions.py:339:        if isinstance(sort, str):
sympy/polys/polyoptions.py:719:        if isinstance(method, str):
sympy/polys/ring_series.py:837:    if isinstance(f, str):
sympy/polys/specialpolys.py:146:    if isinstance(X, str):
sympy/polys/specialpolys.py:149:    if isinstance(Y, str):
sympy/polys/tests/test_polytools.py:82:            return isinstance(a, Poly) and a.eq(b, strict=True)
sympy/printing/codeprinter.py:367:            if isinstance(cond_func, str):
sympy/printing/glsl.py:197:            if isinstance(cond_func, str):
sympy/printing/pretty/stringpict.py:336:        if isinstance(o, str):
sympy/printing/preview.py:180:    if isinstance(expr, str):
sympy/printing/repr.py:33:        if isinstance(expr, str):
sympy/printing/rust.py:303:            if isinstance(cond_func, str):
sympy/printing/str.py:38:        if isinstance(expr, str):
sympy/solvers/ode.py:4366:        if not isinstance(i, str) and i >= 0:
sympy/solvers/ode.py:4482:        if not isinstance(i, str) and i >= 0:
sympy/tensor/tensor.py:1269:    if isinstance(s, str):
sympy/tensor/tensor.py:1505:        if isinstance(s, str):
sympy/utilities/_compilation/runners.py:70:        if isinstance(sources, str):
sympy/utilities/exceptions.py:170:            if not isinstance(value, str):
sympy/utilities/iterables.py:2099:    is_str = isinstance(seq, str)
sympy/utilities/lambdify.py:684:    if isinstance(modules, (dict, str)) or not hasattr(modules, '__iter__'):
sympy/utilities/lambdify.py:859:        if isinstance(args, str):
sympy/utilities/lambdify.py:927:        if isinstance(args, str):
sympy/utilities/lambdify.py:934:        if isinstance(expr, str):
sympy/utilities/lambdify.py:1007:            return isinstance(ident, str) and ident.isidentifier() \
sympy/utilities/lambdify.py:1014:            return isinstance(ident, str) and cls._safe_ident_re.match(ident) \
sympy/utilities/pytest.py:87:        elif isinstance(code, str):
sympy/utilities/runtests.py:504:    if isinstance(kw, str):
sympy/utilities/source.py:26:    if isinstance(lookup_view, str):
```
and
```
sympy/core/numbers.py:53:    if type(z2) is str:
sympy/core/numbers.py:63:            if type(z2) is str and getattr(z1, 'is_Number', False):
sympy/core/numbers.py:1030:            if type(num[1]) is str:
sympy/geometry/util.py:30:    xs = [i for i in free if (i.name if type(x) is str else i) == x]
sympy/ntheory/factor_.py:124:    if type(n) is str:
sympy/utilities/iterables.py:1328:    if len(multiset) == 1 and type(multiset) is str:
sympy/utilities/misc.py:418:        if type(table) is str and type(s) is str:
```

In most cases it should be possible to simply change `str` to `string_types` (and import it), but one will have to look at the context. (Also `is str` must be changed to use `isinstance` for the latter examples.)

### PR DESCRIPTION
**Pull request by divyanshu132**
#### References to other Issues or PRs
Fixes #16094

#### Brief description of what is fixed or changed

fix the unicode support in ` python-2 ` by using ` string_types ` in place of ` str `
#### Release Notes

*  core
   * improve compatibility for unicode strings

### CODE REVIEW

**Comment by divyanshu132** [comment]
@oscargus please review.

File: sympy/polys/orderings.py [hunk 1]
```diff
@@ -225,7 +225,7 @@ def monomial_key(order=None, gens=None):
     if isinstance(order, Symbol):
         order = str(order)
 
-    if isinstance(order, str):
+    if isinstance(order, string_types):
```
Comment (jksuom) [inline sympy/polys/orderings.py]: This is not necessary. All accepted order names (as in `__all__` above) only contain ascii letters.

File: sympy/polys/polyoptions.py [hunk 2]
```diff
@@ -309,7 +309,7 @@ class Wrt(with_metaclass(OptionType, Option)):
     def preprocess(cls, wrt):
         if isinstance(wrt, Basic):
             return [str(wrt)]
-        elif isinstance(wrt, str):
+        elif isinstance(wrt, string_types):
```
Comment (jksuom) [inline sympy/polys/polyoptions.py]: Option names contain ascii letters only.

File: sympy/utilities/misc.py [hunk 3]
```diff
@@ -415,7 +415,7 @@ def translate(s, a, b=None, c=None):
         s = replace(s, mr)
         table = maketrans(a, b)
         # s may have become unicode which uses the py3 syntax for translate
-        if type(table) is str and type(s) is str:
```
Comment (smichr) [inline sympy/utilities/misc.py]: the inline here indicates that the code knows about the unicode possibility and the code should not be changed

**Review by supreet11agrawal — CHANGES_REQUESTED** [review]
Maybe you should add a few test cases as well

**Comment by divyanshu132** [comment]
> Maybe you should add a few test cases as well

I don't think we need tests for this.

File: bin/sympy_time_cache.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 from __future__ import print_function
+from sympy.core.compatibility import string_types
```
Comment (asmeurer) [inline bin/sympy_time_cache.py]: Revert the change to this file. It should not import SymPy before the import at the bottom. I believe the instance where `str` is used can only be a str.

File: doc/ext/docscrape_sphinx.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 from __future__ import division, absolute_import, print_function
+from sympy.core.compatibility import string_types
```
Comment (asmeurer) [inline doc/ext/docscrape_sphinx.py]: Revert this. This file should not import SymPy.

**Comment by asmeurer** [comment]
Tests would be nice where possible. Also this should have a release notes entry.

**Comment by divyanshu132** [comment]
@smichr @asmeurer please review.

File: sympy/core/tests/test_logic.py [hunk 6]
```diff
@@ -172,3 +173,7 @@ def test_formatting():
     raises(ValueError, lambda: S('a&b'))
     raises(ValueError, lambda: S('a|b'))
     raises(ValueError, lambda: S('! a'))
+
+
+def test_issue_16094():
```
Comment (jksuom) [inline sympy/core/tests/test_logic.py]: One substitution test will suffice, and that could be in test_basic.

Comment (divyanshu132) [reply sympy/core/tests/test_logic.py]: so, should I replace all the tests with a single test in test_basic?

Comment (jksuom) [reply sympy/core/tests/test_logic.py]: I think that test_basic would be the proper place for a `subs` test. There is no need for multiple copies in different files. They only repeat the same test.

Comment (oscargus) [reply sympy/core/tests/test_logic.py]: @divyanshu132 If you can figure out what variables that now can be unicode, feel free to add a test to that method where unicode characters/strings are used. It may not always be obvious though.

**Comment by jksuom** [comment]
Thanks, this seems to cover the vulnerable points.
